### DISCUSSION TITLE
Activate/Deactivate thread requirements?

### DISCUSSION THREAD

**Discussion by sagantech** [free-audio/clap#179] [Q&A]
I am having this problem. The plug-in must be deactivated from the audio thread. This is fine for all circumstances except when the host is connected to external audio (say Bluetooth Speaker) and the machine is put to sleep (or possibly unplugged). Under windows in this circumstance it is likely the audio device will be detached completely from the system and no further audio callbacks will be made since the audio device is now gone. How is the host to tell the plug-in to deactivate when there is no longer an audio thread?

**Comment by baconpaul** [comment]
https://github.com/free-audio/clap/blob/af1e05ce6487151c6b36003fdb47566dbb50194e/include/clap/plugin.h#L54

I see `activate` and `deactivate` as `main-thread` not `audio-thread` here.

so while I haven't actually thought about the changing audio buffer state in a host myself, i read that as in your case, when the audio thread goes away, you don't call `stop_processing` ever on the `audio-thread` but you do call `deactivate` on the `main-thread`.

am i misreading the spec or misunderstanding your problem?

**Reply by sagantech** [reply]
Yes, sorry I think stop_processing is what will never be called. So is it ok to call deactivate without stop_processing first?

**Reply by baconpaul** [reply]
That is a question I don’t know. But since there is no requirement to call stop processing in the spec a plugin should survive it.

As Robert says below you could spin up another thread, have thread check return true for is audio thread, and call stop processing from there if you want. But I think that’s probably overkill

**Comment by robbert-vdh** [comment]
Also, a host can have multiple audio threads, and in a non-realtime host the main thread can also be an audio thread. The audio thread precondition is more related to realtime safety than anything else. If you implement the `thread-check` extension you should of course make sure it still returns `true` for the audio thread check even when you use another thread as the plugin's audio thread.

**Reply by sagantech** [reply]
Unfortunately when I skip stop_processing and go directly to Deactivate, U-HE MFM terminates the process with this: 

> [CLAP] Host called clap_plugin.start_processing() twice. 

Works fine with Surge and probably others.

**Reply by alex-tee** [reply]
I'm having the same issue with Stochas for example when I skip stop_processing() and then go to activate the plugin again (e.g., when the block size changes). Seems like hacking the thread-check mechanism to return true temporarily is the only solution unless someone has a better idea.

**Reply by defiantnerd** [reply]
Don't skip stop_processing(). It is part of the state machine of CLAP and a required call.

A lot of plugins use the clap-helpers that detects bad behavior of hosts. 

This is what you are experiencing.
